### PR TITLE
feat(profiling) allocation profiling for PHP ZTS

### PR DESCRIPTION
**Pull request by realFlowControl**
### Description

PROF-8922 / #2070 

This PR will bring allocation profiling to ZTS versions of PHP. In the "there is another custom handler installed" code path we are using an `UnsafeCell` for a thread local which is the cheapest way (evaluated at compile time and not runtime), as this is in the hot path (every allocation and free would borrow check at runtime).
We could argue that the case of another extension that has custom handlers installed is nearly non existing, so we could also be better safe then sorry, but OTOH I could not get it to crash.

### Reviewer checklist
- [x] Test coverage seems ok.
- [x] Appropriate labels assigned.

### CODE REVIEW

**Comment by realFlowControl** [comment]
@morrisonlevi So far I was not able to create crashes due to the`UnsafeCell`, but it feels wrong. Whats your opinion on it?

**Review by morrisonlevi — COMMENTED** [review]
I believe `OnceCell` with `cell.get_or_init(|| unreachable!())` and re-creating the `OnceCell` on each request has good enough codegen to be able to use, and then we avoid the unsafe. Discussed this with Florian on Zoom and Slack.

File: profiling/src/allocation.rs [hunk 1]
```diff
@@ -368,12 +368,18 @@ unsafe extern "C" fn alloc_profiling_malloc(len: size_t) -> *mut c_void {
     ptr
 }
 
+/// Safety: this function pointer is only allowed to point to `allocation_profiling_prev_alloc()`
+/// when at the same time the `ZEND_MM_STATE.prev_custom_mm_alloc` is initialised to a valid
+/// function pointer, otherwise there will be dragons.
 static mut ALLOCATION_PROFILING_ALLOC: unsafe fn(size_t) -> *mut c_void =
```
Comment (morrisonlevi) [inline profiling/src/allocation.rs:374]: In ZTS this needs to be a thread-local, I think?

File: profiling/src/allocation.rs [hunk 2]
```diff
@@ -145,47 +170,50 @@ pub fn allocation_profiling_rinit() {
...
-            ALLOCATION_PROFILING_PREPARE_ZEND_HEAP = prepare_zend_heap;
-            ALLOCATION_PROFILING_RESTORE_ZEND_HEAP = restore_zend_heap;
+
+        if !is_zend_mm() {
+            // Neighboring custom memory handlers found
+            debug!("Found another extension using the ZendMM custom handler hook");
+            unsafe {
+                zend::zend_mm_get_custom_handlers(
+                    // Safety: `unwrap()` is safe here, as `heap` is initialized just above
+                    (*zend_mm_state).heap.unwrap(),
```
Comment (morrisonlevi) [inline profiling/src/allocation.rs:185]: Could probably just save the un-optioned heap from above and avoid the pointer deref + unwrap? Same goes for the `ddog_php_prof_zend_mm_set_custom_handlers` section below.

Comment (realFlowControl) [reply profiling/src/allocation.rs]: Done, I also added more docs

File: profiling/src/allocation.rs [hunk 3]
```diff
@@ -180,36 +178,34 @@ pub fn allocation_profiling_rinit() {
...
         let zend_mm_state = cell.get();
 
-        // Safety: `zend_mm_get_heap()` always returns a non null pointer to a valid heap structure
+        // Safety: `zend_mm_get_heap()` always returns a non-null pointer to a valid heap structure
         let heap = unsafe { Some(zend::zend_mm_get_heap()).unwrap() };
 
-        unsafe {
-            (*zend_mm_state).heap = Some(heap);
-        }
+        unsafe { ptr::addr_of_mut!((*zend_mm_state).heap).write(Some(heap)) };
```
Comment (realFlowControl) [inline profiling/src/allocation.rs]: Just for my understanding: usafe of `ptr::addr_of_mut` removes the creation of an intermediate reference, is that the only reason why you use this?

File: profiling/src/allocation.rs [hunk 4]
```diff
@@ -218,13 +214,13 @@ pub fn allocation_profiling_rinit() {
             zend::ddog_php_prof_zend_mm_set_custom_handlers(
                 heap,
                 Some(alloc_profiling_malloc),
-                Some(alloc_profiling_free),
-                Some(alloc_profiling_realloc),
+                Some(alloc_prof_free),
+                Some(alloc_prof_realloc),
```
[the remark this answers is not in the document]
Comment (realFlowControl) [reply profiling/src/allocation.rs]: My bad, I was going through the commits on a per commit basis and you changed that later 😉

File: profiling/src/allocation.rs (hunk 3 above)
Comment (morrisonlevi) [reply profiling/src/allocation.rs]: Yes. I can't run this particular thing under MIRI, though, I so I can't be _sure_ it's necessary. Avoiding references to the thread-local means there's less chance we screw up the aliasing rules with our unsafe.

**Review by morrisonlevi — APPROVED** [review]

**Pull request merged**